tools/entity-parser.py

```diff
@@ -0,0 +1,152 @@
+                prev_index = index
+                index = src.find("}", index) + 1
+                struct += src[prev_index:index]
+                brace_depth -= 1
+            break
+        if src.find("}", index) < src.find("{", index):
+            prev_index = index
+            index = src.find("}", index) + 1
+            struct += src[prev_index:index]
+            brace_depth -= 1
+        else:
+            prev_index = index
+            index = src.find("{", index) + 1
+            struct += src[prev_index:index]
+            brace_depth += 1
+    return struct + ";"
+
+
+def find_structs(paths):
+    """Return a (name: Struct)-dictionary of all structs found in some files."""
+    expr = re.compile(r"(class|struct)\s+(\w*)(?:\s*:\s*(?:(public|protected|private)\s+(\w*)|(\w*)))?\s*{",
+                      re.ASCII | re.MULTILINE)
+    # (class|struct)
+    # \s+
+    # (\w*)
+    # (?:
+    #   \s*
+    #   :
+    #   \s*
+    #   (?:
+    #       (public|protected|private)
+    #       \s+
+    #       (\w*)
+    #     |
+    #       (\w*)
+    #   )
+    # )?
+    # \s*
+    # {
+
+    # each match contains ( entire match,
+    #                       class/struct
+    #                       name,
+    #                       access modifier,
+    #                       parent name if access modifier is present,
+    #                       parent name if no access modifier is present
+    #                     )
+    # e.g.: [ (...), ('struct', 'A', 'public', 'Entity', ''),
+    #         (...), ('struct', 'E', 'private', 'F', ''),
+    #         (...), ('struct', 'B', '', '', 'Entity'),
+    #         (...), ('struct', 'C', '', '', 'B')
+    #       ]
+
+    structs = {}
+    for path in paths:
+        with open(path, "r") as f:
+            source = process_source("".join(f.readlines()))
+            last_struct_end = 0
+            for match in expr.finditer(source):
+                if not match[2]:  # anonymous struct
+                    continue
+                if match.start() < last_struct_end:  # struct is inside already checked struct
+                    continue
+                if match[3]:  # access modifier present
+                    name, parent = match[2], match[4]
+                else:
+                    name, parent = match[2], match[5]
+                structs[name] = Struct(name, parent)
+                structs[name].source = find_struct_source(source, match.start())
+                last_struct_end = match.start() + len(structs[name].source)
+
+    left = set(structs.keys())
+    while left:
+        struct = structs[left.pop()]
+        if struct.parent in structs:
+            struct.parent = structs[struct.parent]
+        else:
+            struct.parent = None
+    return structs
+
+
+if __name__ == "__main__":
+    lexer = lark.Lark.open("tools/struct.lark", start="structs")
+    for name, struct in find_structs(glob.glob("src/**/*.*", recursive=True)).items():
```

Comment:
You can search for .cpp and .h files. This way we can have documentation in other files without python reading them.

maybe `globi` and an `itertools.chain`?